File: cpp/650_只有两个键的键盘_medium.cpp
class Solution {
public:
    int minSteps(int n) {
        vector<int> dp(n+1);
        for (int i=2; i<n+1; i++) {
            dp[i] = INT_MAX;
            for (int j=1; j*j <= i; j++) {
                int mod = i % j;
                if (mod == 0) {
                    int x = i / j;
                    dp[i] = min(dp[i], dp[j]+x);
                    dp[i] = min(dp[i], dp[x]+j);
                }
            }
        }
        return dp[n];
    }
};

File: python/650_只有两个键的键盘_medium.py
class Solution:
    def minSteps(self, n: int) -> int:
        dp = [0] * (n+1)
        for i in range(2, n+1):
            dp[i] = float("+inf")
            j = 1
            while j*j <= i:
                mod = i % j
                if mod == 0:
                    x = i // j
                    dp[i] = min([dp[i], dp[j]+x, dp[x]+j])
                j += 1
        return dp[n]